### PR TITLE
[Exposing ALTS Context 1/2] Fill in context on TSI and Security Connector Layer

### PR DESCRIPTION
**Pull request by ZhenLian**
This is the first PR to expose ALTS context for users. In includes changes in:

1. alts_tsi_handshaker.cc: convert the fields from alts handshaker result to a grpc_gcp_AltsContext, serialize it into a string, and put it into tsi_peer
2. alts_security_connector.cc: put the new field from tsi_peer to grpc_auth_context
3. corresponding tests

The next PR would be a util function for users to convert from the serialized context to a ProtocolBuffer-generated AltsContext class

### CODE REVIEW

**Comment by ZhenLian** [comment]
None of the failed tests seems to be relevant with the newly added code.
@jiangtaoli2016 @yihuazhang Could you please take a look at this PR?

File: src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc [hunk 1]
```diff
@@ -231,13 +267,33 @@ tsi_result alts_tsi_handshaker_result_create(grpc_gcp_HandshakerResp* resp,
...
   if (!serialized) {
     gpr_log(GPR_ERROR, "Failed to serialize peer's RPC protocol versions.");
     return TSI_FAILED_PRECONDITION;
   }
+  upb::Arena context_arena;
+  grpc_gcp_AltsContext* context = grpc_gcp_AltsContext_new(context_arena.ptr());
+  grpc_gcp_AltsContext_set_application_protocol(context, application_protocol);
+  grpc_gcp_AltsContext_set_record_protocol(context, record_protocol);
+  grpc_gcp_AltsContext_set_security_level(context, 2);
+  grpc_gcp_AltsContext_set_peer_service_account(context, service_account);
```
Comment (yihuazhang) [inline src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc]: Could you please rename it to `peer_service_account` to make it distinguishable from `local_service account`?

File: src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc [hunk 2]
```diff
@@ -231,13 +267,33 @@ tsi_result alts_tsi_handshaker_result_create(grpc_gcp_HandshakerResp* resp,
...
+      peer_rpc_version, rpc_protocol_arena.ptr(), &result->rpc_versions);
   if (!serialized) {
     gpr_log(GPR_ERROR, "Failed to serialize peer's RPC protocol versions.");
     return TSI_FAILED_PRECONDITION;
   }
+  upb::Arena context_arena;
+  grpc_gcp_AltsContext* context = grpc_gcp_AltsContext_new(context_arena.ptr());
+  grpc_gcp_AltsContext_set_application_protocol(context, application_protocol);
+  grpc_gcp_AltsContext_set_record_protocol(context, record_protocol);
+  grpc_gcp_AltsContext_set_security_level(context, 2);
```
Comment (yihuazhang) [inline src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc]: Please add comments about what security level `2` indicates.

File: src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc [hunk 3]
```diff
@@ -231,13 +267,33 @@ tsi_result alts_tsi_handshaker_result_create(grpc_gcp_HandshakerResp* resp,
   result->peer_identity =
       static_cast<char*>(gpr_zalloc(service_account.size + 1));
   memcpy(result->peer_identity, service_account.data, service_account.size);
-  upb::Arena arena;
+  upb::Arena rpc_protocol_arena;
```
Comment (yihuazhang) [inline src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc]: Please rename it to `rpc_versions_arena`.

File: src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc [hunk 4]
```diff
@@ -231,13 +267,33 @@ tsi_result alts_tsi_handshaker_result_create(grpc_gcp_HandshakerResp* resp,
...
+  size_t serialized_ctx_length;
+  char* serialized_ctx = grpc_gcp_AltsContext_serialize(
+      context, context_arena.ptr(), &serialized_ctx_length);
+  if (serialized_ctx == nullptr) {
+    gpr_log(GPR_ERROR, "Failed to serialize peer's ALTS context.");
+    return TSI_FAILED_PRECONDITION;
+  }
+  result->serialized_context =
+      grpc_slice_from_copied_buffer(serialized_ctx, serialized_ctx_length);
+  // gpr_free(serialized_ctx);
```
Comment (yihuazhang) [inline src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc]: Why do not we need to free `serialized_ctx`?

File: src/core/tsi/alts/handshaker/alts_tsi_handshaker.h [hunk 5]
```diff
@@ -28,13 +28,15 @@
 #include "src/core/tsi/alts/handshaker/alts_handshaker_client.h"
 #include "src/core/tsi/transport_security.h"
 #include "src/core/tsi/transport_security_interface.h"
+#include "src/proto/grpc/gcp/altscontext.upb.h"
 #include "src/proto/grpc/gcp/handshaker.upb.h"
 
-#define TSI_ALTS_SERVICE_ACCOUNT_PEER_PROPERTY "service_accont"
```
Comment (yihuazhang) [inline src/core/tsi/alts/handshaker/alts_tsi_handshaker.h]: Thanks for making this change!

File: test/core/tsi/alts/handshaker/alts_tsi_handshaker_test.cc [hunk 6]
```diff
@@ -261,6 +287,25 @@ static void on_client_next_success_cb(tsi_result status, void* user_data,
...
+  upb_strview peer_account = grpc_gcp_AltsContext_peer_service_account(ctx);
+  upb_strview local_account = grpc_gcp_AltsContext_local_service_account(ctx);
+  GPR_ASSERT(memcmp(ALTS_TSI_HANDSHAKER_TEST_APPLICATION_PROTOCOL,
+                    application_protocol.data, application_protocol.size) == 0);
+  GPR_ASSERT(memcmp(ALTS_TSI_HANDSHAKER_TEST_RECORD_PROTOCOL,
+                    record_protocol.data, record_protocol.size) == 0);
+  GPR_ASSERT(memcmp(ALTS_TSI_HANDSHAKER_TEST_PEER_IDENTITY, peer_account.data,
+                    peer_account.size) == 0);
+  GPR_ASSERT(memcmp(ALTS_TSI_HANDSHAKER_TEST_LOCAL_IDENTITY, local_account.data,
+                    local_account.size) == 0);
```
Comment (yihuazhang) [inline test/core/tsi/alts/handshaker/alts_tsi_handshaker_test.cc]: Please add a similar test to `on_server_next_success_cb`.

File: src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc (hunk 4 above)
Comment (ZhenLian) [reply src/core/tsi/alts/handshaker/alts_tsi_handshaker.cc]: Thanks for pointing this out. Just got Confirmed from gRPC team that upb is using arena, which doesn't need to be manually freed.

**Comment by ZhenLian** [comment]
Thanks a lot for the review, Yihua! All comments are resolved. Can you please take another look? @yihuazhang 
Hi @veblush Can you please also take a look at this PR at your earliest convenience? Thanks in advance!

**Pull request merged**